dali/kernels/slice/slice_flip_normalize_permute_cpu.h

```diff
@@ -0,0 +1,272 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SLICE_SLICE_FLIP_NORMALIZE_PERMUTE_CPU_H_
+#define DALI_KERNELS_SLICE_SLICE_FLIP_NORMALIZE_PERMUTE_CPU_H_
+
+#include <utility>
+#include <vector>
+#include "dali/core/common.h"
+#include "dali/core/convert.h"
+#include "dali/core/error_handling.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/slice/slice_kernel_utils.h"
+#include "dali/util/half.hpp"
+
+namespace dali {
+namespace kernels {
+
+namespace detail {
+
+struct ClampPolicy {
+  template <typename OutputType, typename InputType>
+  static inline void Fill(OutputType &destination, InputType element,
+                          const float *mean, const float *inv_stddev) {
+    (void) mean;
+    (void) inv_stddev;
+    destination = clamp<OutputType>(element);
+  }
+};
+
+struct NormalizePolicy {
+  template <typename OutputType, typename InputType>
+  static inline void Fill(OutputType &destination, InputType element,
+                          const float *mean, const float *inv_stddev) {
+    destination = clamp<OutputType>(
+      (static_cast<float>(element) - (*mean)) * (*inv_stddev));
+  }
+};
+
+struct ZeroPadPolicy {
+  template <typename OutputType, typename InputType>
+  static inline void Fill(OutputType &destination, InputType element,
+                          const float *mean, const float *inv_stddev) {
+    (void)element;
+    (void)mean;
+    (void)inv_stddev;
+    destination = static_cast<OutputType>(0);
+  }
+};
+
+template <typename Policy, typename OutputType, typename InputType, size_t Dims>
+void SliceFlipNormalizePermuteImpl(OutputType *output, const InputType *input,
+                                   const std::array<int64_t, Dims> &in_strides,
+                                   const std::array<int64_t, Dims> &out_strides,
+                                   const TensorShape<static_cast<int>(Dims)> &out_shape,
+                                   const TensorShape<static_cast<int>(Dims)> &padded_out_shape,
+                                   const float *mean, const float *inv_stddev,
+                                   size_t normalization_dim,
+                                   const int64_t *permuted_dims,
+                                   std::integral_constant<size_t, 1>) {
+  constexpr auto d = Dims - 1;
+  for (int i = 0; i < out_shape[d]; i++) {
+    size_t norm_idx = normalization_dim == d ? i : 0;
+    Policy::Fill(*output, *input, mean + norm_idx, inv_stddev + norm_idx);
+    input += in_strides[d];
+    output += out_strides[d];
+  }
+
+  for (int i = out_shape[d]; i < padded_out_shape[d]; i++) {
+    size_t norm_idx = normalization_dim == d ? i : 0;
+    ZeroPadPolicy::Fill(*output, *input, mean + norm_idx, inv_stddev + norm_idx);
```

Comment:
This can make static analysis tools scream - and for good reason. `*input` is formally out of bounds already.

Reply:
adjusted as agreed